### PR TITLE
Don't error when exception_check is set to True when return type is PyObject.

### LINKED ISSUE #2529: (@cython.cfunc + Type Annotation) Fragility
Going through some of pandas' cython modules to try to make code valid-python where possible, I'm seeing cythonize-time errors.  It seems like some combinations are invalid for surprising reasons, e.g:

```
cpdef object get_rule_month(object source, object default='DEC')
# existing, works as expected

cpdef object get_rule_month(source: object, default: object='DEC'):
# seems OK

cpdef get_rule_month(source: object, default: object='DEC') -> object:
# breaks --> Syntax error in C variable declaration

@cython.ccall
def get_rule_month(source: object, default: object='DEC') -> object:
# breaks --> Exception clause not allowed for function returning Python object

@cython.ccall
def object get_rule_month(source: object, default: object='DEC'):
# breaks --> Expected '(', found 'get_rule_month'

@cython.ccall
@cython.returns(object)
def get_rule_month(source: object, default: object='DEC'):
# appears OK
```

Another example with a slightly different set of errors:

```
cdef inline bint is_null_datetime64(v):
# existing, OK

cdef inline is_null_datetime64(v) -> bint:
# breaks --> Syntax error in C variable declaration

@cython.cfunc
def inline is_null_datetime64(v) -> bint:
# breaks --> Expected '(', found 'is_null_datetime64

@cython.cfunc
@cython.inline
def is_null_datetime64(v) -> bint:
# breaks --> Function signature does not match previous declaration
```

This last one here is the one that is noticeably different from the first example.

Is this behavior intentional?

### PR DESCRIPTION
**Pull request by matusvalo**
This PR fixes issues in pure python mode when Python object is
annotated as returned value - e.g.:
```python
@cython.cfunc
def foo(s) -> str:
```
It raises following exception: `Exception clause not allowed for function returning Python object`.
This exception is raised because Pure python mode defaults `exception_check` to `True`:

https://github.com/cython/cython/blob/09cbf492dfcf1ad022e79660d6aec0be3f444244/Cython/Compiler/ParseTreeTransforms.py#L2508-L2510

I have tried to fix the issue in this place by checking of returned type but I have found it challenging because
during this stage of compilation we don't have annotation parsed. Hence, I decided to ignore the error and just reset
the `exception_check` value because:

Cython language is ignoring `exception_check` when python object is
returned - e.g. following code is compiled successfully:
```cython
cdef object ftang(object i) except *:
    return 5
```

Hence the pure python mode will have the same behavior as cython language. Moreover I was not able to raise the error in cython language.

Fixes #2529 (If I read comments in the issue correctly all other cases were or fixed or other issues were created)

### CODE REVIEW

**Review by 0dminnimda — COMMENTED** [review]
1)  As stated in #4280 we want `cdef` / `cpdef` to propagate exceptions by defaut and not otherwise for `@cfunc` / `@ccall` 
2) If I understand correctly, with this pr cython it will overwrite the `exception_check` behavior even in case if we specified it explicitly and that's not the best user experience
```
@cython.cfunc
@exceptval(check=True)
def foo(s) -> str:
```

**Comment by matusvalo** [comment]
> As stated in #4280 we want cdef / cpdef to propagate exceptions by default and not otherwise for @cfunc / @ccall

This PR is still keeping this. Exceptions are propagated by default. It only do not error out when object is returned.

> If I understand correctly, with this pr cython it will overwrite the exception_check behaviour even in case if we specified it explicitly and that's not the best user experience

As mentioned in description of PR, this behaviour is present in cython language:

```cython
cdef object ftang(object i) except *:
```

example above in cython is equivalent to your example and still is compiled correctly with no error (hence the `except *` is ignored. So this PR is unifying pure python and cython semantics.

File: Cython/Compiler/Nodes.py [hunk 1]
```diff
@@ -715,7 +715,8 @@ def analyse(self, return_type, env, nonempty=0, directive_locals=None, visibilit
         if (return_type.is_pyobject
                 and (self.exception_value or self.exception_check)
                 and self.exception_check != '+'):
-            error(self.pos, "Exception clause not allowed for function returning Python object")
+            # Exception clause is silently ignored for functions returning Python object.
+            self.exception_check = False
```
Comment (da-woods) [inline Cython/Compiler/Nodes.py:719]: Maybe we should only allow `except *` for functions returning PyObjects (which hopefully matches what gets added with the `cfunc` decorator). Thus _other_ exception statements (e.g. `except -1`) would still be an error?

File: tests/run/pure_py3.py [hunk 2]
```diff
@@ -85,3 +85,12 @@ def call_cdef_inline(x):
...
     return ret, cython.typeof(ret)
+
+@cython.cfunc
+def test_cdef_return_object(x: object) -> object:
+    """
+    Test support of python object in annotations
+    >>> test_cdef_return_object(3)
+    3
+    """
+    return x
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply tests/run/pure_py3.py]: Thanks, done.

File: Cython/Compiler/Nodes.py (hunk 1 above)
Comment (matusvalo) [reply Cython/Compiler/Nodes.py]: Done. I would like to implement test for this case:
```python
@cython.exceptval(-1)
@cython.cfunc
def test_cdef_return_object_broken(x: object) -> object:
    """
    >>> test_cdef_return_object_broken(3)
    # Here I should somehow assert that error() was called.
    """
    return x
```
but I am not sure how I can assert following error:
```
=== Got: ===
105:0: Exception clause not allowed for function returning Python object
```

Comment (da-woods) [reply Cython/Compiler/Nodes.py]: The files to check for compile-time errors are in tests/errors. If you follow one of them it'll show you how to do it

Comment (matusvalo) [reply Cython/Compiler/Nodes.py]: Done.

**Comment by matusvalo** [comment]
Based on message of last commit I have checked whether we can avoid setting `self.exception_check` to `False`:
```python
        elif return_type.is_pyobject and self.exception_check:
            pass

        if (return_type.is_pyobject
                and (self.exception_value)
                and self.exception_check != '+'):
            error(self.pos, "Exception clause not allowed for function returning Python object")
```

Unfortunately, with this change Cython compiler crashes for `pure_py3` test:
```
ERROR: runTest (__main__.CythonRunTestCase)
compiling (c/cy2) and running pure_py3
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ness-skmac3/dev/cython/runtests.py", line 1449, in run
    ext_so_path = self.runCompileTest()
  File "/Users/ness-skmac3/dev/cython/runtests.py", line 1092, in runCompileTest
    return self.compile(
  File "/Users/ness-skmac3/dev/cython/runtests.py", line 1370, in compile
    self._match_output(expected_errors, errors, tostderr)
  File "/Users/ness-skmac3/dev/cython/runtests.py", line 1422, in _match_output
    self.assertEqual(None, unexpected)
AssertionError: None != '89:0: Compiler crash in AnalyseDeclarationsTransform'
```

Hence, I suppose we need to leave it there. @scoder is there anything in this PR to be done?

**Comment by da-woods** [comment]
You may be able to get to the bottom of the compiler crash by compiling it outside the test runner. Usually `python3 cythonize.py tests/run/pure_py.py` (maybe with `-2` as well since most of the tests use this by default). That should give you a more extended traceback.

I don't think there's a good reason to remove the `self.exception_check = False` though. I think the message from the last commit was about avoiding testing `self.exception_check` with the `is` operator rather than removing the assignment.

**Pull request merged**